cluster-autoscaler/core/scale_down.go

```diff
@@ -103,6 +103,93 @@ func (n *NodeDeleteStatus) SetDeleteInProgress(status bool) {
 	n.deleteInProgress = status
 }
 
+type resourceName string
+type scaleDownResourcesLimits map[resourceName]int64
+type scaleDownResourcesDelta map[resourceName]int64
+
+func computeScaleDownResourcesLeftLimits(nodes []*apiv1.Node, cp cloudprovider.CloudProvider, timestamp time.Time) (scaleDownResourcesLimits, errors.AutoscalerError) {
+	resourceLimiter, err := cp.GetResourceLimiter()
+	if err != nil {
+		return scaleDownResourcesLimits{}, errors.ToAutoscalerError(
+			errors.CloudProviderError,
+			err)
+	}
+
+	totals := make(map[resourceName]int64)
+	totals[cloudprovider.ResourceNameCores], totals[cloudprovider.ResourceNameMemory] = calculateCoresAndMemoryTotal(nodes, timestamp)
+
+	for gpuName, gpuCount := range calculateGpusTotal(nodes, cp, timestamp) {
+		totals[resourceName(gpuName)] = gpuCount
+	}
+
+	resultScaleDownLimits := make(scaleDownResourcesLimits)
+	for _, resource := range resourceLimiter.GetResources() {
```

Comment:
I know it's nitpicking, but we call a method here that creates a set of all resources for which either min or max limit is given and returns their names, only to retrieve only non-zero mins for them anyway. What about:

```for resource, min := range resourceLimiter.GetMinResources() {```

where `GetMinResources` would simply return the map with min resources from ResourceLimiter?

Reply:
Then I would need to extend the interface with `GetMinResources`. Which is less general. Also I feel that it is better to have methods which do just one thing (separate method for getting resource names and separate for getting min/max value for a resource). Not a method which does two things together.

Reply:
> Then I would need to extend the interface with `GetMinResources`. 
I don't think we have a Go interface for it, it's just one struct type shared across cloud providers. And `GetResources` is only used here, so it's trading one method for another. 

>  Which is less general. 
Do we have a use case for this (supposedly) more general method?

> Also I feel that it is better to have methods which do just one thing (separate method for getting resource names and separate for getting min/max value for a resource). Not a method which does two things together.

GetMinResources (a simple getter with a one-line implementation):
- returns minResources field.

GetResources (a method that extracts just the names of all resources for which either min or max limit was defined):
- creates a map of structs to act like a set,
- adds resources from maxResources to it,
- adds resources from minResources to it,
- copies the keys from it to a slice.

It's kind of a stretch to say that the former does "two things" and more than the latter. In fact, we could even get rid of it entirely if the ResourceLimiter's fields were public.

Reply:
> I don't think we have a Go interface for it, it's just one struct type shared across cloud providers. And GetResources is only used here, so it's trading one method for another.

I did not mean "golang" maning of word "interface". Just test of methods defined on `ResourceLimiter`.

> Do we have a use case for this (supposedly) more general method?

It is just a good principle to have methods which have one responsibility. The `GetMinResources` method you are proposing contradicts that. And we are not winning anything at all IMO (just less readable code - at least to me).

> It's kind of a stretch to say that the former does "two things" and more than the latter.

The inner implementation is irrelevant. By "thing" I mean responsiblity. And given that:
 * `getResources` just lists resource names for which limits are defined (one thing)
 * `getMinResources` lists resource names for which min limit is defined AND lists values of those limits (two things).